File: find_music.cpp
#include <iostream>
#include <string>
#include <getopt.h>
#include <fstream>
#include <filesystem>
#include <map>

using namespace std;
namespace fs = std::filesystem;

string temp_folder = "temp";
string extension = "";

int compress_and_get_size(string compressor, string file);
string concat_signatures(string compressor, string file1, string file2);
float ncd(int x_size, int y_size, int xy_size);
void set_extension(string compressor);
string get_file_name(string file);

int main(int argc, char* argv[]) {

    string compressor = "gzip";
    string sample = "";
    string freqs_folder = "";

    if (argc < 5) {
        cerr << "Usage: find_music [ -c compressor ] [ -s sample ] [ -f freqs_folder ]" << endl;
        return 1;
    }

    int opt;
    while ((opt = getopt(argc, argv, "c:s:f:")) != -1) {
        switch (opt) {
            // compressors: gzip, bzip2, lzma, zstd, xz, lzop, lz4
            case 'c':
                if (string(optarg) != "gzip" && string(optarg) != "bzip2" && string(optarg) != "lzma" && string(optarg) != "zstd" && string(optarg) != "xz" && string(optarg) != "lzop" && string(optarg) != "lz4") {
                    cerr << "Error: invalid compressor" << endl;
                    return 1;
                }
                compressor = optarg;
                break;
            case 's':
                sample = optarg;
                break;
            case 'f':
                freqs_folder = optarg;
                break;
            default:
                cerr << "Usage: find_music [ -c compressor ] [ -s sample ] [ -f freqs_folder ]" << endl;
                return 1;
        }
    }

    // Check if the compressor is available
    if (system(("which " + compressor + " > /dev/null").c_str()) != 0) {
        cerr << "Error: " << compressor << " is not available" << endl;
        return 1;
    }

    set_extension(compressor);

    string extra_args = "";
    // read args from freqs_folder args file
    ifstream args_file(freqs_folder + "/args");
    if (args_file.is_open()) {
        string line;
        while (getline(args_file, line)) {
            extra_args = line;
        }
        args_file.close();
    }

    if (!fs::exists(temp_folder)) {
        fs::create_directory(temp_folder);
    }

    // get the signature of the sample
    string command = "./GetMaxFreqs/bin/GetMaxFreqs -w \"" + temp_folder + "/sample.freqs\"" + extra_args + " '" + sample + "'";
    if (system(command.c_str()) != 0) {
        cerr << "Error: could not generate the signature of the sample" << endl;
        return 1;
    }

    // compress the signature and get the size of the compressed file
    int sample_size = compress_and_get_size(compressor, temp_folder + "/sample.freqs");

    map<float, string> music_ncd;

    string music_signature_path = "";
    string music_name = "";
    int music_signature_size = 0;
    string concat_file = "";
    int concat_size = 0;
    for (const auto & entry : fs::directory_iterator(freqs_folder)) {
        if (entry.path().extension() != ".freq") {
            continue;
        }
        music_signature_path = entry.path();
        music_signature_size = compress_and_get_size(compressor, music_signature_path);
        if (music_signature_size == -1) {
            cerr << "Error: could not compress the signature of the music" << endl;
            return 1;
        }

        concat_file = concat_signatures(compressor, temp_folder + "/sample.freqs", music_signature_path);
        if (concat_file == "") {
            cerr << "Error: could not concatenate the signature of the sample and the music" << endl;
            return 1;
        }
        concat_size = compress_and_get_size(compressor, concat_file);
        if (concat_size == -1) {
            cerr << "Error: could not compress the concatenated signature of the sample and the music" << endl;
            return 1;
        }

        music_name = get_file_name(music_signature_path);
        music_ncd[ncd(sample_size, music_signature_size, concat_size)] = music_name;
    }

    for (auto it = music_ncd.rbegin(); it != music_ncd.rend(); ++it) {
        cout << it->first << " : " << it->second << endl;
    }

    fs::remove_all(temp_folder);

    return 0;
}

int compress_and_get_size(string compressor, string file) {
    string file_name = get_file_name(file);
    string command = compressor + " -c \'" + file + "\' > \'" + temp_folder + "/" + file_name + "." + extension + "\'";
    if (system(command.c_str()) != 0) {
        return -1;
    }

    ifstream sample_file(temp_folder + "/" + file_name + "." + extension);
    if (!sample_file.is_open()) {
        return -1;
    }

    sample_file.seekg(0, ios::end);
    int sample_size = sample_file.tellg();
    sample_file.close();

    return sample_size;
}

string concat_signatures(string compressor, string file1, string file2) {
    string file1_name = get_file_name(file1);
    string file2_name = get_file_name(file2);

    string command = "cat \'" + file1 + "\' \'" + file2 + "\' | " + compressor + " > \'" + temp_folder + "/" + file1_name + "_" + file2_name + ".freq" + "\'";
    if (system(command.c_str()) != 0) {
        return "";
    }

    return temp_folder + "/" + file1_name + "_" + file2_name + ".freq";
}

float ncd(int x_size, int y_size, int xy_size) {
    return (float)(xy_size - min(x_size, y_size)) / max(x_size, y_size);
}

void set_extension(string compressor) {
    if (compressor == "gzip") {
        extension = "gz";
    } else if (compressor == "bzip2") {
        extension = "bz2";
    } else if (compressor == "lzma") {
        extension = "lzma";
    } else if (compressor == "zstd") {
        extension = "zst";
    } else if (compressor == "xz") {
        extension = "xz";
    } else if (compressor == "lzop") {
        extension = "lzo";
    } else if (compressor == "lz4") {
        extension = "lz4";
    }
}

string get_file_name(string file) {
    string file_name = file.substr(file.find_last_of("/\\") + 1);
    return file_name.substr(0, file_name.find_last_of("."));
}

File: group_2/src/find_music.py
from argparse import ArgumentParser
import os
import shutil

TEMP_FOLDER = 'temp'
extension = ''

def compress_and_get_size(compressor: str, compressor_args: str, file: str) -> int:
    file_name = file.split('/')[-1]

    command = f'{compressor} {compressor_args} -c -k -f "{file}" > "{TEMP_FOLDER}/{file_name}.{extension}"'
    os.system(command)

    return os.path.getsize(f'{TEMP_FOLDER}/{file_name}.{extension}')

def concat_signatures(compressor: str, file1: str, file2: str) -> None:
    file1_name = file1.split('/')[-1]
    file2_name = file2.split('/')[-1]

    command = f'cat "{file1}" "{file2}" | {compressor} -c -k -f > "{TEMP_FOLDER}/{file1_name}_{file2_name}.freq"'
    os.system(command)

    return f'{TEMP_FOLDER}/{file1_name}_{file2_name}.freq'

def ncd(x_size: int, y_size: int, xy_size: int) -> float:
    return (xy_size - min(x_size, y_size)) / max(x_size, y_size)

def set_extension(compressor: str) -> None:
    global extension
    if compressor == "gzip":
        extension = "gz"
    if compressor == "bzip2":
        extension = "bz2"
    if compressor == "lzma":
        extension = "lzma"
    if compressor == "zstd":
        extension = "zst"
    if compressor == "xz":
        extension = "xz"
    if compressor == "lzop":
        extension = "lzo"

def find_music(compressor: str, compressor_args: str, sample: str, freqs_folder: str) -> list:

    set_extension(compressor)

    music_args = ""
    with open(f'{freqs_folder}/__args', 'r') as f:
        music_args = f.read().strip()

    if not os.path.exists(TEMP_FOLDER):
        os.mkdir(TEMP_FOLDER)

    command = f'./GetMaxFreqs/bin/GetMaxFreqs -w "{TEMP_FOLDER}/sample.freq" {music_args} "{sample}"'
    os.system(command)

    sample_size = compress_and_get_size(compressor, compressor_args, f'{TEMP_FOLDER}/sample.freq')

    music_ncd = {}

    for file in os.listdir(freqs_folder):
        if file.endswith('.freq'):
            file_size = compress_and_get_size(compressor, compressor_args, f'{freqs_folder}/{file}')
            concat_file = concat_signatures(compressor, f'{TEMP_FOLDER}/sample.freq', f'{freqs_folder}/{file}')
            concat_size = compress_and_get_size(compressor, compressor_args, concat_file)
            music_name = file.split('.')[0]
            music_ncd[music_name] = ncd(sample_size, file_size, concat_size)

    shutil.rmtree(TEMP_FOLDER)

    sorted_music_ncd = [(k, v) for k, v in sorted(music_ncd.items(), key=lambda item: item[1])]

    return sorted_music_ncd

def main():
    
    parser = ArgumentParser(description='Find the most similar audio files using Normalized Compression Distance')
    parser.add_argument('-c', '--compressor', help='Compressor to use (gzip, bzip2, lzma, zstd, xz, lzop)', type=str, required=True)
    parser.add_argument('-s', '--sample', help='Sample file', type=str, required=True)
    parser.add_argument('-f', '--freqs_folder', help='Folder with frequency files', type=str, required=True)
    args = parser.parse_args()

    if not os.path.exists(args.freqs_folder):
        print('Folder with frequency files does not exist')
        return
    
    if not os.path.exists(args.sample):
        print('Sample file does not exist')
        return
    
    compressor = args.compressor.split()
    compressor_args = " ".join(compressor[1:])
    compressor = compressor[0]

    if compressor not in ['gzip', 'bzip2', 'lzma', 'zstd', 'xz', 'lzop']:
        print('Invalid compressor')
        return
    
    if os.system(f'which {compressor} > /dev/null') != 0:
        print(f'{compressor} is not installed or not in PATH')
        return

    ranking = find_music(compressor, compressor_args, args.sample, args.freqs_folder)

    for music, ncd_value in reversed(ranking):
        print("{:.6f} | {}".format(ncd_value, music))
    print(f'\nBest match:\n\t{ranking[0][0]}\n')
    
    
if __name__ == '__main__':
    main()